src/gpuarray/array.h

```diff
@@ -607,6 +607,78 @@ GPUARRAY_PUBLIC void GpuArray_fprintf(FILE *fd, const GpuArray *a);
 
 GPUARRAY_PUBLIC int GpuArray_fdump(FILE *fd, const GpuArray *a);
 
+/************************************************************************************
+*                         multi-gpu collectives interface                          *
+************************************************************************************/
+
+/**
+ * \brief TODO
+ * \param src [const GpuArray*] TODO
+ * \param dest [GpuArray*] TODO
+ * \param count [int] TODO
+ * \param typecode [int] TODO
+ * \param opcode [int] TODO
+ * \param root [int] TODO
+ * \param comm [gpucomm*] TODO
+ * \return int TODO
+ */
+GPUARRAY_PUBLIC int GpuArray_reduce(const GpuArray* src, GpuArray* dest,
+                                    int count, int typecode, int opcode,
+                                    int root, gpucomm* comm);
```

Comment:
I am trying to think of a array-level wrapper, that would not include the privately defined `gpucomm`. But I think we will have to cope with some unnecessary difficulties:

A `GpuArray` holds a single `gpudata` pointer and "metadata". A `gpudata` instance holds a single `gpucontext`. So from any `GpuArray` you can access the `gpucontext`. But a gpu can be associated with multiple `gpucomm`s, because a user could want to have a group of gpus to execute a certain operation together and a different group to execute another on the same `GpuArray`. For example, one could define a grid of gpus with their respective `GpuArray`s and would like first to concurrently execute an AllReduce sum operation on the columns and then to concurrently execute an AllReduce product operation on the rows so that we can get - let's say - the Πi(Σj arrays[i, j]) on every gpu. And I do not believe that the user can spare the multiple `gpucomm`s per gpu in any implementation of this or that we should restrict them to use one `gpucomm` per gpu. So, because I believe that we cannot associate a `gpucontext` with a dynamic array of `gpucomm`s without providing a "clumsy"** interface for the user to act on a desired `gpucomm` in array-level, I think that we should expose `gpucomm` to the array-level and let the user manage where he thinks he should use it. As an argument in favor of this, nccl or mpi library would not expose the communication worlds if there could be a more user-friendly way without restricting the possibilities.  

** By "clumsy" I mean that an interface of `GpuArray_all_reduce`, which would include an integer to have user select one of the already registered `gpucomm` within a `gpucontext`, would not be well-defined.